File: day16_offline_parallel.py
import sys


with open("day16.txt", "r") as f:
    data = f.read()
    
lines = data.split("\n")
pattern = []
for line in lines:
    row = []
    for char in line:
        row.append(char)
    pattern.append(row)
    
def get_next_beams(beam_row, beam_col, direction, char):
    if char == '.':
        # keep going
        if direction == "right":
            return [(beam_row, beam_col + 1, direction)]
        elif direction == "left":
            return [(beam_row, beam_col - 1, direction)]
        elif direction == "up":
            return [(beam_row - 1, beam_col, direction)]
        elif direction == "down":
            return [(beam_row + 1, beam_col, direction)]
    elif char == '/':
        # need do reflect
        if direction == "right":
            return [(beam_row - 1, beam_col, "up")]
        elif direction == "left":
            return [(beam_row + 1, beam_col, "down")]
        elif direction == "up":
            return [(beam_row, beam_col + 1, "right")]
        elif direction == "down":
            return [(beam_row, beam_col - 1, "left")]
    elif char == '\\':
        # also reflect
        if direction == "right":
            return [(beam_row + 1, beam_col, "down")]
        elif direction == "left":
            return [(beam_row - 1, beam_col, "up")]
        elif direction == "up":
            return [(beam_row, beam_col - 1, "left")]
        elif direction == "down":
            return [(beam_row, beam_col + 1, "right")]
    elif char == '|':
        # point side: keep going
        # flat side: split
        if direction == "right" or direction == "left":
            return [(beam_row + 1, beam_col, "down"), (beam_row - 1, beam_col, "up")]
        elif direction == 'up':
            return [(beam_row - 1, beam_col, "up")]
        elif direction == 'down':
            return [(beam_row + 1, beam_col, "down")]
    elif char == '-':
        if direction == "up" or direction == "down":
            return [(beam_row, beam_col + 1, "right"), (beam_row, beam_col - 1, "left")]
        elif direction == 'right':
            return [(beam_row, beam_col + 1, "right")]
        elif direction == 'left':
            return [(beam_row, beam_col - 1, "left")]
    raise Exception("Unknown char: " + char)

def get_next_position(beam_row, beam_col, direction):
    if direction == "right":
        return (beam_row, beam_col + 1)
    elif direction == "left":
        return (beam_row, beam_col - 1)
    elif direction == "up":
        return (beam_row - 1, beam_col)
    elif direction == "down":
        return (beam_row + 1, beam_col)

def count_number_of_non_zero(energized):
    count = 0
    for row in range(len(energized)):
        for col in range(len(energized[row])):
            if energized[row][col] != 0:
                count += 1
    return count


edge_start_beams = []
# top
if sys.argv[1] == "0":
    for idx in range(len(pattern[0])):
        edge_start_beams.append((0, idx, "down"))
# bottom
elif sys.argv[1] == "1":
    for idx in range(len(pattern[-1])):
        edge_start_beams.append((len(pattern) - 1, idx, "up"))
# left
elif sys.argv[1] == "2":
    for idx in range(len(pattern)):
        edge_start_beams.append((idx, 0, "right"))
# right
elif sys.argv[1] == "3":
    for idx in range(len(pattern)):
        edge_start_beams.append((idx, len(pattern[0]) - 1, "left"))
    

edge_energized = []

for outer_idx, edge_start_beam in enumerate(edge_start_beams):
    print("calculating edge: ", outer_idx, edge_start_beam)
    energized = []
    for row in range(len(pattern)):
        energized_row = []
        for col in range(len(pattern[row])):
            energized_row.append(0)
        energized.append(energized_row)

    beams = [edge_start_beam]
    traced_beams = []
    sim_step = 0
    current_energized = 0

    while len(beams) > 0 and sim_step <= 100000:
        sim_step += 1
        last_energized = current_energized
        current_beam = beams.pop(0)
        beam_row, beam_col, direction = current_beam
        if sim_step % 1000 == 0:
            print("step: " + str(sim_step) + " beam: " + str(current_beam) + " direction: " + direction, "len beam", len(beams))
        # print("step: " + str(sim_step) + " beam: " + str(current_beam) + " direction: " + direction, "len beam", len(beams))
        # out of border, ignore
        if beam_row < 0 or beam_row >= len(pattern):
            # print("out of border", beam_row, beam_col, direction)
            continue
        if beam_col < 0 or beam_col >= len(pattern[beam_row]):
            # print("out of border", beam_row, beam_col, direction)
            continue
        
        energized[beam_row][beam_col] += 1

        next_char = pattern[beam_row][beam_col]
        next_beams = get_next_beams(beam_row, beam_col, direction, next_char)
        # print("next beams: " + str(next_beams))

        for next_beam in next_beams:
            if next_beam in traced_beams:
                continue
            else:
                traced_beams.append(next_beam)
                beams.append(next_beam)

        current_energized = count_number_of_non_zero(energized)
        # print("current energized: " + str(current_energized), "last energized: " + str(last_energized))
    
    edge_energized.append(current_energized)
    print("calculating edge: ", outer_idx, edge_start_beam, "done, energized=", current_energized)

print(edge_energized)
print("max", max(edge_energized))

File: day16_offline_parallel.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <tuple>
#include <algorithm>


using namespace std;

vector<vector<char>> readPatternFromFile(const string& filePath) {
    vector<vector<char>> pattern;
    ifstream file(filePath);
    if (file.is_open()) {
        string line;
        while (getline(file, line)) {
            if (line.empty()) {
                continue; // Ignore empty lines
            }
            vector<char> row;
            for (char c : line) {
                if (c != '\n') {
                    row.push_back(c);
                }
            }
            pattern.push_back(row);
        }
        file.close();
    }
    return pattern;
}

vector<tuple<int, int, string>> getNextBeams(int beamRow, int beamCol, const string& direction, char c) {
    vector<tuple<int, int, string>> nextBeams;

    if (c == '.') {
        if (direction == "right") {
            nextBeams.push_back(make_tuple(beamRow, beamCol + 1, direction));
        } else if (direction == "left") {
            nextBeams.push_back(make_tuple(beamRow, beamCol - 1, direction));
        } else if (direction == "up") {
            nextBeams.push_back(make_tuple(beamRow - 1, beamCol, direction));
        } else if (direction == "down") {
            nextBeams.push_back(make_tuple(beamRow + 1, beamCol, direction));
        }
    } else if (c == '/') {
        if (direction == "right") {
            nextBeams.push_back(make_tuple(beamRow - 1, beamCol, "up"));
        } else if (direction == "left") {
            nextBeams.push_back(make_tuple(beamRow + 1, beamCol, "down"));
        } else if (direction == "up") {
            nextBeams.push_back(make_tuple(beamRow, beamCol + 1, "right"));
        } else if (direction == "down") {
            nextBeams.push_back(make_tuple(beamRow, beamCol - 1, "left"));
        }
    } else if (c == '\\') {
        if (direction == "right") {
            nextBeams.push_back(make_tuple(beamRow + 1, beamCol, "down"));
        } else if (direction == "left") {
            nextBeams.push_back(make_tuple(beamRow - 1, beamCol, "up"));
        } else if (direction == "up") {
            nextBeams.push_back(make_tuple(beamRow, beamCol - 1, "left"));
        } else if (direction == "down") {
            nextBeams.push_back(make_tuple(beamRow, beamCol + 1, "right"));
        }
    } else if (c == '|') {
        if (direction == "right" || direction == "left") {
            nextBeams.push_back(make_tuple(beamRow + 1, beamCol, "down"));
            nextBeams.push_back(make_tuple(beamRow - 1, beamCol, "up"));
        } else if (direction == "up") {
            nextBeams.push_back(make_tuple(beamRow - 1, beamCol, "up"));
        } else if (direction == "down") {
            nextBeams.push_back(make_tuple(beamRow + 1, beamCol, "down"));
        }
    } else if (c == '-') {
        if (direction == "up" || direction == "down") {
            nextBeams.push_back(make_tuple(beamRow, beamCol + 1, "right"));
            nextBeams.push_back(make_tuple(beamRow, beamCol - 1, "left"));
        } else if (direction == "right") {
            nextBeams.push_back(make_tuple(beamRow, beamCol + 1, "right"));
        } else if (direction == "left") {
            nextBeams.push_back(make_tuple(beamRow, beamCol - 1, "left"));
        }
    }
    return nextBeams;
}

tuple<int, int> getNextPosition(int beamRow, int beamCol, const string& direction) {
    if (direction == "right") {
        return make_tuple(beamRow, beamCol + 1);
    } else if (direction == "left") {
        return make_tuple(beamRow, beamCol - 1);
    } else if (direction == "up") {
        return make_tuple(beamRow - 1, beamCol);
    } else if (direction == "down") {
        return make_tuple(beamRow + 1, beamCol);
    }
}

int countNumberOfNonZero(const vector<vector<int>>& energized) {
    int count = 0;
    for (const auto& row : energized) {
        for (int value : row) {
            if (value != 0) {
                count++;
            }
        }
    }
    return count;
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        cout << "Usage: ./program <file_path> <edge_index>" << endl;
        return 1;
    }

    string filePath = argv[1];
    int edgeIndex = stoi(argv[2]);

    vector<vector<char>> pattern = readPatternFromFile(filePath);

    cout << "Shape of pattern: " << pattern.size() << " rows x " << pattern[0].size() << " columns" << endl;


    vector<tuple<int, int, string>> edgeStartBeams;
    if (edgeIndex == 0) {
        for (int idx = 0; idx < pattern[0].size(); idx++) {
            edgeStartBeams.push_back(make_tuple(0, idx, "down"));
        }
    } else if (edgeIndex == 1) {
        for (int idx = 0; idx < pattern.back().size(); idx++) {
            edgeStartBeams.push_back(make_tuple(pattern.size() - 1, idx, "up"));
        }
    } else if (edgeIndex == 2) {
        for (int idx = 0; idx < pattern.size(); idx++) {
            edgeStartBeams.push_back(make_tuple(idx, 0, "right"));
        }
    } else if (edgeIndex == 3) {
        for (int idx = 0; idx < pattern.size(); idx++) {
            edgeStartBeams.push_back(make_tuple(idx, pattern[0].size() - 1, "left"));
        }
    }

    vector<int> edgeEnergized;

    for (int outerIdx = 0; outerIdx < edgeStartBeams.size(); outerIdx++) {
        cout << "calculating edge: " << outerIdx << " " << get<0>(edgeStartBeams[outerIdx]) << " " << get<1>(edgeStartBeams[outerIdx]) << endl;
        vector<vector<int>> energized(pattern.size(), vector<int>(pattern[0].size(), 0));

        vector<tuple<int, int, string>> beams = {edgeStartBeams[outerIdx]};
        vector<tuple<int, int, string>> tracedBeams;
        int simStep = 0;
        int currentEnergized = 0;

        while (!beams.empty() && simStep <= 100000) {
            simStep++;
            int lastEnergized = currentEnergized;
            tuple<int, int, string> currentBeam = beams.front();
            beams.erase(beams.begin());
            int beamRow = get<0>(currentBeam);
            int beamCol = get<1>(currentBeam);
            string direction = get<2>(currentBeam);

            if (simStep % 1000 == 0) {
                cout << "step: " << simStep << " beam: " << beamRow << " " << beamCol << " direction: " << direction << " len beam: " << beams.size() << endl;
            }

            if (beamRow < 0 || beamRow >= pattern.size() || beamCol < 0 || beamCol >= pattern[beamRow].size()) {
                continue;
            }

            energized[beamRow][beamCol]++;

            char nextChar = pattern[beamRow][beamCol];
            vector<tuple<int, int, string>> nextBeams = getNextBeams(beamRow, beamCol, direction, nextChar);

            for (const auto& nextBeam : nextBeams) {
                if (find(tracedBeams.begin(), tracedBeams.end(), nextBeam) != tracedBeams.end()) {
                    continue;
                } else {
                    tracedBeams.push_back(nextBeam);
                    beams.push_back(nextBeam);
                }
            }

            currentEnergized = countNumberOfNonZero(energized);
        }

        edgeEnergized.push_back(currentEnergized);
        cout << "calculating edge: " << outerIdx << " " << get<0>(edgeStartBeams[outerIdx]) << " " << get<1>(edgeStartBeams[outerIdx]) << " done, energized=" << currentEnergized << endl;
    }

    int maxEnergized = *max_element(edgeEnergized.begin(), edgeEnergized.end());
    cout << "edge energized: ";
    for (int energized : edgeEnergized) {
        cout << energized << " ";
    }
    cout << endl;
    cout << "max: " << maxEnergized << endl;

    return 0;
}
